cmd/xl-zones.go

```diff
@@ -0,0 +1,645 @@
+package cmd
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"sort"
+
+	"github.com/minio/minio/pkg/lifecycle"
+	"github.com/minio/minio/pkg/madmin"
+	"github.com/minio/minio/pkg/policy"
+	"github.com/minio/minio/pkg/sync/errgroup"
+)
+
+type xlZones struct {
+	zones []*xlSets
+}
+
+// Initialize new zone of erasure codes.
+func newXLZones(endpointsList EndpointLists, formats []*formatXLV3) (ObjectLayer, error) {
+	z := &xlZones{}
+	for i, ep := range endpointsList {
+		sets, err := newXLSets(ep.Endpoints, formats[i], ep.SetCount, ep.DrivesPerSet)
+		if err != nil {
+			return nil, err
+		}
+		z.zones = append(z.zones, sets)
+	}
+	return z, nil
+}
+
+// AvailableZoneSlice attaches the methods of Uint64erface to []uint64, sorting in increasing order.
```

Comment:
Simplify (and fix comments):

```Go
// AvailableZoneIndex contains the zone index and availability.
type AvailableZoneIndex struct {
	Index     int
	Available uint64
}

// AvailableZoneSlice is a slice of available zones.
type AvailableZoneSlice []AvailableZoneIndex

// getLeastUsageZoneSlice returns the available zones 
// ordered by disk availability, least to most.
func (z *xlZones) getLeastUsageZoneSlice(ctx context.Context) AvailableZoneSlice {
	var availableLayers AvailableZoneSlice
	for i, objLayer := range z.zones {
		info := objLayer.StorageInfo(ctx)
		var available uint64
		for _, davailable := range info.Available {
			available += davailable
		}
		availableLayers = append(availableLayers, struct {
			Index     int
			Available uint64
		}{
			Index:     i,
			Available: available,
		})
	}
	sort.Slice(availableLayers, func(i, j int) bool {
		return availableLayers[i].Available < availableLayers[j].Available
	})
	return availableLayers
}
```

See note below on whether the comparison should be flipped.